### PR TITLE
Bigquery script create

### PR DESCRIPTION
**Pull request by JSchoenbachler**
Adds bigquery functionality via R script to backup from an existing PostgreSQL DB

### CODE REVIEW

File: R/drop_pmid_version.R [hunk 1]
```diff
@@ -38,8 +38,14 @@ deleteOldPmidVersions = function(tableSuffix, dryRun, dbtype, dbname, ...) {
     DBI::dbRemoveTable(con, tableKeep)
   } else {
     DBI::dbRemoveTable(con, tableNow)
-    q = glue_sql('alter table {`tableKeep`} rename to {`tableNow`}', .con = con)
-    n = DBI::dbExecute(con, q)}
+    if(inherits(con, 'BigQueryConnection')){
```
Comment (jakejh) [inline R/drop_pmid_version.R:41]: Clean up whitespace in places like this.

File: R/get_parsing_tables.R [hunk 2]
```diff
@@ -98,13 +98,17 @@ getParsingTables = function(tableSuffix) {
 
 
 createParsingTables = function(
-  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL,
+  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL, tabNames = c(),
```
Comment (jakejh) [inline R/get_parsing_tables.R:101]: - Change tabNames to tableNames for consistency.
- use default of NULL, and change the if statement accordingly

File: R/get_parsing_tables.R [hunk 3]
```diff
@@ -98,13 +98,17 @@ getParsingTables = function(tableSuffix) {
...
   ...) {
   if (is.null(dbname)) return(invisible())
 
   con = connect(dbtype, dbname, ...)
   parTables = getParsingTables(tableSuffix)
 
+  if(!is.na(tabNames) && length(tabNames > 0)){
+    parTables = parTables[tabNames == names(parTables)]
```
Comment (jakejh) [inline R/get_parsing_tables.R:109]: Use `names(parTables) %in% tableNames` inside the brackets. The current code will not work if length(tabNames) > 1.

File: R/utilities.R [hunk 4]
```diff
@@ -185,20 +189,26 @@ getPgParams = function(path = '~/.pgpass') {
...
+  # get total number of lines
+  n = R.utils::countLines(path)
+
+  # read first row to get data types, but don't send it to database
+  d = data.table::fread(path, nrows = 1L)
+
   # will be inefficient if file is compressed
-  if (DBI::dbExistsTable(con, tableName)) {
+  if (DBI::dbExistsTable(con, tableName) && isTRUE(append)) {
+  } else if(DBI::dbExistsTable(con, tableName)) {
```
Comment (jakejh) [inline R/utilities.R:201]: Remove redundancy in nested if statements.

File: R/utilities.R [hunk 5]
```diff
@@ -185,20 +189,26 @@ getPgParams = function(path = '~/.pgpass') {
   x3}
 
 
-writeTableInChunks = function(path, con, nRowsPerChunk, overwrite, tableName) {
+writeTableInChunks = function(path, con, nRowsPerChunk, overwrite, tableName, append) {
```
Comment (jakejh) [inline R/utilities.R:192]: Use `overwrite` or `append` but not both.

File: data-raw/create-clickhouse-from-postgres.R [hunk 6]
```diff
@@ -0,0 +1,46 @@
+createClickhouseFromPostgres = function(pgDbName = 'pmdbclick', clDbName = 'pmdbclick', chunkSize = 15000L){
```
Comment (jakejh) [inline data-raw/create-clickhouse-from-postgres.R:1]: Hyphenated file seems redundant with underscore file.

**Review by jakejh — CHANGES_REQUESTED** [review]
Remove missing_abstract*.csv files.

File: R/drop_pmid_version.R (hunk 1 above)
Comment (JSchoenbachler) [reply R/drop_pmid_version.R:41]: Could you explain how you want the whitespace to be?

**Comment by JSchoenbachler** [comment]
All items other than the ones responded to have been addressed in latest pull

**Comment by JSchoenbachler** [comment]
So a potential issue....bigquery has API limits. I saw I had hit this after running on a particularly large set of data:

https://cloud.google.com/bigquery/quotas#query_jobs

Essentially, we can only update a table 1,500 times per day. I was running at 5k chunkSize, meaning we would have a 7.5 mil max per day. This also doesn't include any retries, which could be an issue (and have been with some values getting dropped off). I'm looking into a potential workaround. Maybe if the code recognizes there would be more than 1.5k updates in a call, it splits into separate tables we can then combine? There has to be a simpler solution. I'm looking into it.

**Comment by jakejh** [comment]
That's another good reason to make the chunkSize table-specific. For most
tables, I think we could set the chunkSize way higher, such as 1e5 or even
1e6.

File: R/drop_pmid_version.R (hunk 1 above)
Comment (jakejh) [reply R/drop_pmid_version.R:41]: use `if (...) {`, not `if(...){`

File: data-raw/create_bigquery_from_postgres.R [hunk 7]
```diff
@@ -3,16 +3,22 @@ library(data.table)
 library(glue)
 library(bigrquery)
 
-createBigQueryFromPostgres = function(pgDbName = 'pmdb', project = 'pmparser-test', dataset = 'pmdb', tableName, chunkSize = 15000L, overwriteTable = FALSE, tables = c()){
+#tables variable is a named list where the name is the table name and the value is the chunk size you want to process. Chunk sizes of 0 or NA will be removed from list
```
Comment (jakejh) [inline data-raw/create_bigquery_from_postgres.R:6]: The `tables` variable seems redundant with `chunkSize` and its name is misleading. Figure out how to use only the `chunkSize` argument.

**Review by jakejh — CHANGES_REQUESTED** [review]

**Comment by jakejh** [comment]
Have you tested everything goes smoothly with the latest changes? I'll try to test in the next few days.

**Comment by JSchoenbachler** [comment]
I have tested as going along, but will run another test tomorrow on a small scale, then larger scale after that so I can test everything together.

File: R/get_parsing_tables.R [hunk 8]
```diff
@@ -98,13 +98,17 @@ getParsingTables = function(tableSuffix) {
...
 createParsingTables = function(
-  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL,
+  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL, tableNames = NULL,
   ...) {
   if (is.null(dbname)) return(invisible())
 
   con = connect(dbtype, dbname, ...)
   parTables = getParsingTables(tableSuffix)
 
+  if(!is.null(tableNames) && length(tableNames > 0)){
```
Comment (jakejh) [inline R/get_parsing_tables.R:108]: - whitespace
- Why does the if statement need the `&& length(...`?

**Review by jakejh — CHANGES_REQUESTED** [review]
Don't forget to address the redundancy in the nested if statements in `writeTableInChunks`.

File: R/get_parsing_tables.R (hunk 8 above)
Comment (JSchoenbachler) [reply R/get_parsing_tables.R:108]: I just want to make sure an empty list isn't passed as an extra precaution, but if you don't think it's necessary then I'll just remove it!

File: R/get_parsing_tables.R [hunk 9]
```diff
@@ -88,22 +88,30 @@ getParsingTables = function(tableSuffix) {
...
-  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL,
-  ...) {
+  tableSuffix = NULL, overwrite = FALSE, dbtype = 'postgres', dbname = NULL, ...) {
   if (is.null(dbname)) return(invisible())
 
   con = connect(dbtype, dbname, ...)
-  parTables = getParsingTables(tableSuffix)
+  parTables = getParsingTables(tableSuffix, ...)
+
+  if(dbtype == 'bigquery'){
```
Comment (jakejh) [inline R/get_parsing_tables.R:110]: This code is now unnecessary, correct?

File: R/get_parsing_tables.R [hunk 10]
```diff
@@ -1,4 +1,4 @@
-getParsingTables = function(tableSuffix) {
+getParsingTables = function(tableSuffix, tableNames = NULL, ...) {
```
Comment (jakejh) [inline R/get_parsing_tables.R:1]: I don't see why `getParsingTables` needs the ellipsis.

File: R/utilities.R [hunk 11]
```diff
@@ -186,19 +190,24 @@ getPgParams = function(path = '~/.pgpass') {
...
 writeTableInChunks = function(path, con, nRowsPerChunk, overwrite, tableName) {
+  # get total number of lines
+  n = R.utils::countLines(path)
+
+  # read first row to get data types, but don't send it to database
+  d = data.table::fread(path, nrows = 1L)
+
   # will be inefficient if file is compressed
-  if (DBI::dbExistsTable(con, tableName)) {
+  if (!(DBI::dbExistsTable(con, tableName) && !isTRUE(overwrite))) {
```
Comment (jakejh) [inline R/utilities.R:200]: I don't understand the reasoning behind the complicated logic of the changed code. Why should the function not remove the old table, if it exists and overwrite is true?

**Comment by JSchoenbachler** [comment]
Closing in lieu of the new bigquery script made by jake.

**Pull request closed without merging**